ambari-server/src/main/java/org/apache/ambari/server/topology/validators/NameNodeHaValidator.java

```diff
@@ -0,0 +1,128 @@
+ */
+public class NameNodeHaValidator implements TopologyValidator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(NameNodeHaValidator.class);
+  private static final Splitter SPLITTER = Splitter.on(',').omitEmptyStrings().trimResults();
+
+  @Override
+  public void validate(ClusterTopology topology) throws InvalidTopologyException {
+    if( topology.isNameNodeHAEnabled() ) {
+
+      if (!allHostNamesAreKnown(topology)) {
+        LOG.warn("Provision cluster template contains hostgroup(s) without explicit hostnames. Cannot validate NAMENODE HA setup in this case.");
+        return;
+      }
+
+      Collection<String> nnHosts = topology.getHostAssignmentsForComponent("NAMENODE");
+
+      if (nnHosts.isEmpty()) {
+        LOG.info("NAMENODE HA is enabled but there are no NAMENODE components in the cluster. Assuming external name nodes.");
+        validateExternalNamenodeHa(topology);
+      }
+      else if (nnHosts.size() == 1) {
+        throw new InvalidTopologyException("NAMENODE HA requires at least two hosts running NAMENODE but there is only one: " +
+          nnHosts.iterator().next());
+      }
+
+      Map<String, String> hadoopEnvConfig = topology.getConfiguration().getFullProperties().get("hadoop-env");
+      if(hadoopEnvConfig != null && !hadoopEnvConfig.isEmpty() && hadoopEnvConfig.containsKey("dfs_ha_initial_namenode_active") && hadoopEnvConfig.containsKey("dfs_ha_initial_namenode_standby")) {
+        if((!HostGroup.HOSTGROUP_REGEX.matcher(hadoopEnvConfig.get("dfs_ha_initial_namenode_active")).matches() && !nnHosts.contains(hadoopEnvConfig.get("dfs_ha_initial_namenode_active")))
+          || (!HostGroup.HOSTGROUP_REGEX.matcher(hadoopEnvConfig.get("dfs_ha_initial_namenode_standby")).matches() && !nnHosts.contains(hadoopEnvConfig.get("dfs_ha_initial_namenode_standby")))){
+          throw new InvalidTopologyException("NAMENODE HA hosts mapped incorrectly for properties 'dfs_ha_initial_namenode_active' and 'dfs_ha_initial_namenode_standby'. Expected hosts are: " + nnHosts);
+        }
+      }
+    }
+
+  }
+
+  private boolean allHostNamesAreKnown(ClusterTopology topology) {
+    return topology.getHostGroupInfo().values().stream().allMatch(
+      hg -> !hg.getHostNames().isEmpty());
+  }
+
+  /**
+   * Verifies that all respective properties are set for an external NANENODE HA setup
+   */
+  public void validateExternalNamenodeHa(ClusterTopology topology) throws InvalidTopologyException {
+    Map<String, String> hdfsSite = topology.getConfiguration().getFullProperties().get("hdfs-site");
+    Set<String> hosts = topology.getAllHosts();
+
+    for (Map.Entry<String, List<String>> entry: getNameServicesToNameNodes(hdfsSite).entrySet()) {
+      String nameService = entry.getKey();
+      List<String> nameNodes = entry.getValue();
+      for (String nameNode: nameNodes) {
+        String address = hdfsSite.get("dfs.namenode.rpc-address." + nameService + "." + nameNode);
```

Comment:
Minor issue:  Should we consider adding checks for the other endpoint properties (dfs.namenode.http-address.*, etc)?

Reply:
I don't have a strong opinion. Are the http/https addresses only used if **webhdfs** is enabled or in all cases?

Reply:
I'm not sure when these properties were needed, but I think they may be used in either case (whether or not webhdfs is enabled).  

For now, I think the check you have is fine, but maybe we should check with an HDFS HA expert, and eventually amend this if need be.  

Thanks.